File: 20151214_challenge_245_easy_date_dilemma/java/date-dillema/src/main/java/br/com/ramon/datedillema/DateParser.java
package br.com.ramon.datedillema;

public class DateParser{

	public String parse(String date){
		String[] splitted = splitDate(date);

		String formatedDate = "";

		if (splitted[0].length() <= 2) {
			formatedDate = formatPatternMDY(splitted);
		} else if (splitted[0].length() == 4) {
			formatedDate = formatPatternYMD(splitted);
		}

		return formatedDate;
	}

	private String formatPatternYMD(String[] splittedDate){
		String year = "";
		String month = "";
		String day = "";
		
		year = formateYear(splittedDate[0]);

		month = formateMonth(splittedDate[1]);
		
		day = formateDay(splittedDate[2]);

		return year+"-"+month+"-"+day;
	}

	private String formatPatternMDY(String[] splittedDate){
		String year = "";
		String month = "";
		String day = "";
		month = formateMonth(splittedDate[0]);
		
		day = formateDay(splittedDate[1]);

		year = formateYear(splittedDate[2]);

		return year+"-"+month+"-"+day;
	}

	private String[] splitDate(String date){
		if (date.indexOf('/') > -1) {
			return date.split("/");	
		}

		if (date.indexOf('-') > -1) {
			return date.split("-");
		}

		if (date.indexOf(' ') > -1) {
			return date.split(" ");
		}

		return new String[]{};
	}

	private String formateMonth(String month){
		if (month.length() <= 2) {
			return formatNumberLowerThan10(month);
		}
		return month;
	}

	private String formateDay(String day){
		if (day.length() <= 2) {
			return formatNumberLowerThan10(day);	
		}
		return day;
	}

	private String formateYear(String year){
		if (year.length() <= 2) {
			return "20" + formatNumberLowerThan10(year);	
		}
		return year;		
	}

	private String formatNumberLowerThan10(String number){
		if (number.length() == 1) {
			return "0" + number;
		}
		return number;
	}
}

File: 20151214_challenge_245_easy_date_dilemma/scala/date-dillema/src/main/scala/br/com/ramon/datedillema/DateParser.scala
package br.com.ramon.datedillema

/**
  * Created by ramon on 27/06/16.
  */
class DateParser {
  def parse(input: String): String = {
    val splitedDate = splitDate(input)
    val formatedDate = splitedDate.head.length match {
      case mdy if mdy <= 2 => formatPatternMDY(splitedDate)
      case ymd if ymd == 4 => formatPatternYMD(splitedDate)
    }
    formatedDate
  }

  private def splitDate(date: String): Array[String] = {
    Array("/", "-", " ").foreach { c =>
      if (date.indexOf(c) > -1) return date.split(c)
    }
    new Array[String](0)
  }

  private def formatPatternYMD(splittedDate: Array[String]): String = {
    val year = formateYear(splittedDate(0))

    val month = formateMonth(splittedDate(1))

    val day = formateDay(splittedDate(2))

    year + "-" + month + "-" + day
  }

  private def formatPatternMDY(splittedDate: Array[String]): String = {
    val month = formateMonth(splittedDate(0))

    val day = formateDay(splittedDate(1))

    val year = formateYear(splittedDate(2))
    year + "-" + month + "-" + day
  }


  private def formateMonth(month: String): String = {
    if (month.length() <= 2) {
      return formatNumberLowerThan10(month)
    }
    month
  }

  private def formateDay(day: String): String = {
    if (day.length() <= 2) {
      return formatNumberLowerThan10(day)
    }
    day
  }

  private def formateYear(year: String): String = {
    if (year.length() <= 2) {
      return "20" + formatNumberLowerThan10(year)
    }
    year
  }

  private def formatNumberLowerThan10(number: String): String = {
    if (number.length() == 1) {
      return "0" + number
    }
    number
  }

}
